### PR TITLE
Replace String shard gen with ShardGeneration

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we use Strings for lots of different things when manipulating
snapshots; one crucial such thing is a shard generation. We're not very
consistent about naming the variables containing these things, and have
other kinds of generation in use, so it takes extra effort to track
shard generations through the code. This commit introduces a
`ShardGeneration` class to encapsulate just those strings that are used
as shard generations.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/ShardGeneration.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.Random;
+
+/**
+ * The generation ID of a shard, used to name the shard-level {@code index-$SHARD_GEN} file that represents a {@link
+ * BlobStoreIndexShardSnapshots} instance. Before 7.6 ({@link SnapshotsService#SHARD_GEN_IN_REPO_DATA_VERSION}) these generations were
+ * numeric, but recent versions use a UUID instead.
+ */
+public final class ShardGeneration implements Writeable, ToXContentFragment {
+
+    private final String rawGeneration;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/ShardGeneration.java:33]: I pondered alternatives e.g. a more memory-efficient representation such as just a `byte[]`. I think we don't have too many of these things in play at once, O(1) per shard in the repository, so decided not to bother yet.

**Review by DaveCTurner — COMMENTED** [review]
For discussion, I'm not dead-set on doing this, but I did find it hard to follow how these things flow around the code when they're just strings.

**Review by original-brownbear — APPROVED** [review]
LGTM, I like this a lot. This would also allow e.g. making the "special" generations "_new" and "_delete" a little more obvious in a follow-up I think + a number of additional clarifications.

File: server/src/main/java/org/elasticsearch/repositories/ShardGeneration.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/ShardGeneration.java:32]: ++ the memory use of these been somewhere between trivial and small in practice even in the worst case scenarios where large repos were causing trouble, there's better spots to invest in if we wanted to improve snapshot memory consumption.

**Comment by DaveCTurner** [comment]
Backport pending on #75733 (maybe more, #75501 perhaps)

**Comment by DaveCTurner** [comment]
Actually the conflicts were pretty trivial, backported in 1e8ecbcc3492b9187eaa5c63ecde2efe534f6721.
